### PR TITLE
Fix some item type magic numbers

### PR DESCRIPTION
**Pull request by Dystopian**
**When merged this pull request will:**
- title.

These macros are used: https://github.com/acemod/ACE3/blob/2767fddd92bd48199ba580601dd148f77e001f00/addons/main/script_macros.hpp#L65-L68

### CODE REVIEW

**Comment by SilentSpike** [comment]
Good stuff, way more readable 👍

No idea about those other types personally

**Comment by dedmen** [comment]
Looking into it..
Interesting 🤔 
```
"PrimaryWeapon"
"SecondaryWeapon"
"RightHand"
"LeftHand"
"Backpack"
"Binocular"
"EarPhones"
"Map"
"Compass"
"Watch"
"Radio"
"GPS"
"Para"
"Scuba"
"Goggles"
"HMD"
"MineDetector"
"Medikit"
"Toolkit"
"Backpack2"
"Headgear"
"Vest"
"Hair"
"NVG"
"Launcher"
"Pistol"
"PistolHolstered"
"ForceTracker"
"Handgun"
"FinL"
"FinR"
```
Left and right hand? ForceTracker? Ear phones?

**Comment by dedmen** [comment]
Found it!
If class has no "ItemInfo" and no "WeaponSlotsInfo" entry.
Then `type` is a mask.
```
#define MaskSlotPrimary     0x00000001  // primary weapons
#define MaskSlotHandGun     0x00000002  // hand gun
#define MaskSlotSecondary   0x00000004  // secondary weapons
#define MaskSlotItem        0x00000F00  // items
#define MaskSlotBinocular   0x00003000  // binocular
#define MaskHardMounted     0x00010000  // hard mounted
#define MaskSlotHandGunItem 0x000000F0  // hand gun magazines
#define MaskSlotInventory   0x001E0000  // hand gun magazines
```
But 4096 is missing 🤔 
The first 3 you can see 
https://github.com/acemod/ACE3/blob/2767fddd92bd48199ba580601dd148f77e001f00/addons/arsenal/functions/fnc_scanConfig.sqf#L105-L113

ah here. 
https://github.com/acemod/ACE3/blob/2767fddd92bd48199ba580601dd148f77e001f00/addons/arsenal/functions/fnc_scanConfig.sqf#L74
Binocular is 4096 but.. According to my stuff it's 0x3000 not 0x1000 🤔 

I'd say just make a `#define TYPE_BINOCULAR 4096` and also the primary/secondary/handgun defines for 1,4,2

**Comment by Dystopian** [comment]
> I'd say just make a #define TYPE_BINOCULAR 4096 and also the primary/secondary/handgun defines for 1,4,2

what about these? https://github.com/acemod/ACE3/blob/2767fddd92bd48199ba580601dd148f77e001f00/addons/common/functions/fnc_getItemType.sqf#L45-L47 https://github.com/acemod/ACE3/blob/2767fddd92bd48199ba580601dd148f77e001f00/addons/common/functions/fnc_getItemType.sqf#L69-L74
The second one looks not like binocular only.

**Comment by dedmen** [comment]
`2^4` -> 0x10 -> 16
`2^8` -> 0x100 -> 256
`2^9` -> 0x200 -> 512
`2^12` -> 0x1000 -> 4096
These should also be replaced by the appropriate macros btw. We are doing constant calculations at runtime here.

I'm failing to find the actual 4096 anywhere... Maybe.. The second one is actually only binocular? I think this data just comes from BIS_fnc_itemType. Maybe these items actually don't exist at all. One should run with tons of mods and check if that part of the code is ever executed.


Duh. I found the 4096. Mask for Binocular is 0x3000. That means it's binary AND'ed together.
And 0x1000 & 0x3000 -> 0x1000 -> true
also 0x2000 & 0x3000 -> 0x2000 -> true

and https://github.com/Dystopian/ACE3/blob/ceb544bd3d8f384cfb540f33f02efe52d47cb825/addons/common/functions/fnc_getItemType.sqf#L80
is `MaskSlotInventory` because
`0x20000 & 0x001E0000` -> 0x20000 -> true.

Meh.. complex stuff.

File: addons/arsenal/functions/fnc_addVirtualItems.sqf [hunk 1]
```diff
@@ -81,37 +81,37 @@ if (_items isEqualType true) then {
...
-                                case 302: {
+                                case TYPE_BIPOD: {
                                     (_cargo select 1) select 3 pushBackUnique _x;
                                 };
                             };
                         };
                         /* Headgear */
                         case (isClass (_configItemInfo) &&
-                            {getNumber (_configItemInfo >> "type") == 605}): {
+                            {getNumber (_configItemInfo >> "type") == TYPE_HEADGEAR}): {
```
Comment (commy2) [inline addons/arsenal/functions/fnc_addVirtualItems.sqf]: isEqualTo

File: addons/arsenal/functions/fnc_scanConfig.sqf [hunk 2]
```diff
@@ -46,23 +46,23 @@ private _configCfgWeapons = configFile >> "CfgWeapons"; //Save this lookup in va
...
                 {!(configName _x isKindOf ["CBA_MiscItem", (_configCfgWeapons)])}
             ): {
 
             //Convert type to array index
-            (_cargo select 1) select ([201,301,101,302] find _itemInfoType) pushBackUnique _className;
+            (_cargo select 1) select ([TYPE_OPTICS,TYPE_FLASHLIGHT,TYPE_MUZZLE,TYPE_BIPOD] find _itemInfoType) pushBackUnique _className;
         };
         /* Headgear */
-        case (_itemInfoType == 605): {
+        case (_itemInfoType == TYPE_HEADGEAR): {
```
Comment (commy2) [inline addons/arsenal/functions/fnc_scanConfig.sqf]: isEqualTo

File: addons/common/functions/fnc_getItemType.sqf [hunk 3]
```diff
@@ -47,24 +47,24 @@ switch (true) do {
...
-    case (_type == 605): {["item", "headgear"]};
-    case (_type == 611): {["item", "radio"]};
-    case (_type == 616): {["item", "hmd"]};
-    case (_type == 617): {["item", "binocular"]};
-    case (_type == 619): {["item", "medikit"]};
-    case (_type == 620): {["item", "toolkit"]};
-    case (_type == 621): {["item", "uav_terminal"]};
-    case (_type == 701): {["item", "vest"]};
-    case (_type == 801): {["item", "uniform"]};
+    case (_type == TYPE_MUZZLE): {["item", "muzzle"]};
```
Comment (commy2) [inline addons/common/functions/fnc_getItemType.sqf]: isEqualTo

File: addons/common/functions/fnc_getItemType.sqf [hunk 4]
```diff
@@ -47,24 +47,24 @@ switch (true) do {
...
+    case (_type == TYPE_RADIO): {["item", "radio"]};
+    case (_type == TYPE_HMD): {["item", "hmd"]};
+    case (_type == TYPE_BINOCULAR): {["item", "binocular"]};
+    case (_type == TYPE_MEDIKIT): {["item", "medikit"]};
+    case (_type == TYPE_TOOLKIT): {["item", "toolkit"]};
+    case (_type == TYPE_UAV_TERMINAL): {["item", "uav_terminal"]};
+    case (_type == TYPE_VEST): {["item", "vest"]};
+    case (_type == TYPE_UNIFORM): {["item", "uniform"]};
 
     case (_type == 2^12): {
```
Comment (commy2) [inline addons/common/functions/fnc_getItemType.sqf]: Looks good so far. Please replace these as well.

File: addons/arsenal/functions/fnc_addVirtualItems.sqf (hunk 1 above)
Comment (Dystopian) [reply addons/arsenal/functions/fnc_addVirtualItems.sqf]: these functions can't gain noticeable performance (nanoseconds in comparison to overall microseconds) from this replacement but become really less readable

File: addons/main/script_macros.hpp [hunk 5]
```diff
@@ -60,6 +60,18 @@
...
+#define TYPE_WEAPON_HANDGUN 2
+#define TYPE_WEAPON_SECONDARY 4
+// magazine types
+#define TYPE_MAGAZINE_HANDGUN_AND_GL 16 // mainly
+#define TYPE_MAGAZINE_PRIMARY_AND_THROW 256
+#define TYPE_MAGAZINE_SECONDARY_AND_PUT 512 // mainly
+// more types
+#define TYPE_BINOCULAR_AND_NVG 4096
+#define TYPE_WEAPON_VEHICLE 65536
+#define TYPE_ITEM 131072
```
Comment (dedmen) [inline addons/main/script_macros.hpp]: Engine name `WeaponSlotInventory` Is less helpful though.

File: addons/arsenal/functions/fnc_scanConfig.sqf [hunk 6]
```diff
@@ -142,7 +142,7 @@ private _putList = [];
     switch true do {
         // Rifle, handgun, secondary weapons mags
         case (
-                (getNumber (_x >> "type") in [256,512,1536,16]) &&
+                (getNumber (_x >> "type") in [TYPE_MAGAZINE_PRIMARY_AND_THROW,TYPE_MAGAZINE_SECONDARY_AND_PUT,1536,TYPE_MAGAZINE_HANDGUN_AND_GL]) &&
```
Comment (dedmen) [inline addons/arsenal/functions/fnc_scanConfig.sqf]: Can't find out what 1536 is. It's 0x600 and thus falls under `MaskSlotItem`
Would be interesting to know what items fall under that.

Comment (Dystopian) [reply addons/arsenal/functions/fnc_scanConfig.sqf]: some rockets which are not `TYPE_MAGAZINE_SECONDARY_AND_PUT` IIRC

Comment (dedmen) [reply addons/arsenal/functions/fnc_scanConfig.sqf]: The number behind there defines the `SlotCount`.. Don't really understand what it means though. Certainly not it's size.

Comment (commy2) [reply addons/arsenal/functions/fnc_scanConfig.sqf]: Can confirm that A2 inventory size was changed by the type value.

Comment (commy2) [reply addons/arsenal/functions/fnc_scanConfig.sqf]: E.g. 1x1 mag vs 2x1 mag vs 6x2 mag etc.

Comment (dedmen) [reply addons/arsenal/functions/fnc_scanConfig.sqf]: yeah then that's it. This should really be implemened using binary & then... But we don't have that so meh

**Pull request merged**